tools/ttysnoop.py

```diff
@@ -147,18 +148,27 @@ def usage():
     if (from->type != (ITER_IOVEC + WRITE))
         return 0;
 #else
-    if (from->iter_type != ITER_IOVEC)
+    if (from->iter_type != ITER_UBUF && from->iter_type != ITER_IOVEC)
         return 0;
     if (from->data_source != WRITE)
         return 0;
 #endif
 
-
-    kvec  = from->kvec;
-    buf   = kvec->iov_base;
-    count = kvec->iov_len;
-
-    return do_tty_write(ctx, kvec->iov_base, kvec->iov_len);
+    switch (from->iter_type) {
+    case ITER_IOVEC:
+        kvec  = from->kvec;
+        buf   = kvec->iov_base;
+        count = kvec->iov_len;
+        break;
```

Comment:
So the address space is in kernel space for ITER_IOVEC, and it is in user space for ITER_UBUF. If this is the case, I guess, we should have add a parameter to do_tty_write to differentiate these two cases since the corresponding bpf_probe_read_* functions will be different.

Reply:
Hi, Song. I'm curious about kernel/user space case. Since always, `ttysnoop.py`'s probe of `buf` has `__user` macro. Does it's prove that we should use `bpf_probe_read_user()` in both case. By the way, i tried `bpf_probe_read_kernel()` for `ITER_IOVEC`, it's cause `ttysnoop.py` run failed.

AND, In the new commit just now, I optimized the program for CO-RE.

Thanks a lot.